### ISSUE TITLE
How can I get all requires for any package via code?

### ISSUE THREAD

**Issue by andreybuzin** [composer/composer#4399]
I want to be able to draw a tree of dependencies for any package.

**Comment by alcohol** [comment]
So basically you want to copy what https://github.com/clue/graph-composer does?

**Comment by andreybuzin** [comment]
It seems that this extension (graph-composer) uses composer.lock of ready project. Maybe I'm wrong. My task for any packages without reference to a project.

**Comment by alcohol** [comment]
The concept is still the same. Except with lock file you know specific versions, with composer.json you cannot know version.

**Comment by clue** [comment]
> Except with lock file you know specific versions, with composer.json you cannot know version.

This is the main reason graph-composer relies on your lock file and there's little we can do about this. This applies as along as you care about the actual versions of your dependencies.

The lock file explicitly lists every package among with version information which composer actually found to be compatible.

Let's assume your root package has a dependency on "vendor/A:*"

* "vendor/A:1" has a dependency on "vendor/B:*" 
* "vendor/A:2" has a dependency on "vendor/C:*"

Now which packages are you going to add to the graph? This alone is already non-trivial, but you easily make this more complex by introducing additional constraints such as "vendor/C:1" depends on "root:1" etc.

Long story short, you're probably better off relying on the dependency resolution algorithm composer currently uses and stick with its lock file.

If you want to draw the graph of *any* package, you may want to look into using `composer create-project` and then running `graph-composer` on the resulting directory. Though keep in mind that this still takes your local platform constraints (php version, extensions etc.) into account.

One other alternative would be ignoring the actual package versions and draw a graph of all packages ever referenced. In the above example this would mean that you root package would link to vendor/A which would like to vendor/B and vendor/C.

I hope this helps :)

**Comment by andreybuzin** [comment]
Thanks for the answers, guys. I need a way using (!)only composer.json.
For example, forget at all about the graph. I can create a composer.json file in an empty directory. Then I can run "composer install" and Composer will install each package from a "require" section and dependencies for the package. 
The only question is (!)how in text form I can get the information about all dependencies and relations between dependencies relative to the root (root = selected package by me from "require" section of composer.json). 

"root:1" has a dependency on "vendor/A:*"
"vendor/A:*" has a dependency on "vendor/B:2"
"vendor/B:2" has a dependency on "vendor/C:*"

It seems there is no contradiction with the above.

**Comment by alcohol** [comment]
But as @clue already made clear, running composer install could lead to different results depending on environment context. So there is no way to guarantee your graph will match reality. Composer install result ends up in composer.lock, but the result is specific to the environment you ran the install on. What you are asking for is not really doable realistically. You can do a composer install --dry-run and then parse the outcome using grep/sed etc, but that is just plain nasty. Alternatively, figure out how composer works internally. But be aware that the internal API is not set in stone.

**Comment by andreybuzin** [comment]
Yes, I noticed the words about environment context. But Composer requires PHP 5.3.2+ to run. What else? My last qustion:
Is it possible to call the Composer with 3 parameters? 
1. current package, 
2. current version, 
3. environment context as a variable (for example, ['PHP' => 5.3.2] or empty variable like null).

**Comment by alcohol** [comment]
The requirements of Composer have nothing to do with this.

**Comment by andreybuzin** [comment]
Let my application dependent on environment context. I'll try to do it. It's bad, but better than nothing :)

The issue can be closed, if there are no others suggestions for resolving of this problem.
Thanks.

**Comment by stof** [comment]
> environment context as a variable

you cannot put it in a variable. But you can put it in the composer.json: https://getcomposer.org/doc/06-config.md#platform
